src/UploadQueue.cpp

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (C) 2017 Prism Skylabs
+ */
+#include "private/UploadQueue.h"
+#include "boost/thread/locks.hpp"
+#include "boost/format.hpp"
+#include "easylogging++.h"
+
+namespace
+{
+struct NotEmpty
+{
+    NotEmpty(const std::deque<prism::connect::UploadArtifactTaskPtr>& d)
+        : deque(d)
+    {}
+
+    bool operator()() const { return !deque.empty(); }
+
+    const std::deque<prism::connect::UploadArtifactTaskPtr>& deque;
+};
+}
+
+namespace prism
+{
+namespace connect
+{
+
+// Used internally by push_back. Caller must lock mutex_ before calling.
+bool UploadQueue::arrangeFreeSpaceForTask(const size_t taskSize)
+{
+    if(taskSize > maxMemorySize_)
+        return false;
+
+    while(taskSize + size_ > maxMemorySize_)
+    {
+        const UploadArtifactTaskPtr t = deque_.front();
+        deque_.pop_front();
+        addSize(-(int)t->getArtifactSize());
+        LOG(WARNING) << "Upload queue is full. Preemptively removed " << t->toString();
+    }
+    return true;
+}
+
+void UploadQueue::push_back(UploadArtifactTaskPtr task)
+{
+    const size_t artifactSize = task ? task->getArtifactSize() : 0;
+    bool canPush = false;
+    {
+        boost::lock_guard<boost::mutex> lock(mutex_);
+        canPush = arrangeFreeSpaceForTask(artifactSize);
+        if(canPush)
+        {
+            deque_.push_back(task);
+            addSize(artifactSize);
+        }
+    }
+    cv_.notify_one();
+    if(!canPush)
+    {
+        const std::string message = (boost::format("Artifact %s is too large (%d) to put into upload queue. "
+                "Queue max size is: %d bytes") % task->toString() % artifactSize % maxMemorySize_).str();
+        LOG(ERROR) << message;
```

Comment:
How do you suggest to process this error (in addition to logging) it? We can't put artifact into the queue. That's it. We shouldn't abort application. Because it's embedded device and application is supposed to run for days, weeks, possibly months and more.